protocol messages 3.6 support

petrilakm opened the pull request:
Issue #16.
Proper function F13-F28 handling.
Proper momentary/permanent handling on functions F0-F28.
Working on hand controller LH-100 v3.6

Stupid code copy&paste, need big cleanup.

petrilakm committed:
protocol messages 3.6 support

Proper function F13-F28 handling.
Proper momentary/permanent handling on functions F0-F28.
Working on hand controller LH-100 v3..6

Stupid code copy&paste, need big cleanup.

petrilakm commented on src/tHnaciVozidlo.pas:
```diff
@@ -26,6 +26,7 @@   THVPomCV = record // jeden zaznam POM se sklada z
   end;
 
   THVFuncType = (permanent = 0, momentary = 1);
+  TFunkceType = array [0 .. _MAX_FUNC] of THVFuncType; // typ funkci HV
```
Nový typ pro momentary/permanent nastavení funkcí. Dle vzoru stavu funkcí.

petrilakm commented on src/tHnaciVozidlo.pas:
```diff
@@ -53,7 +54,7 @@   THV = class
     funcVyznam: array [0 .. _MAX_FUNC] of string;
     // seznam popisu funkci hnaciho vozidla
-    funcType: array [0 .. _MAX_FUNC] of THVFuncType;
+    funcType: TFunkceType;
```
Nový typ pro momentary/permanent nastavení funkcí. Dle vzoru stavu funkcí.

petrilakm commented on src/tUltimateLI.pas:
```diff
@@ -113,6 +113,9 @@   TuLI = class
     procedure SendKeepAlive();
     procedure SendStatusRequest();
     procedure SendLocoData(callByte: Byte; addr: Integer);
+    procedure SendLocoFunc13(callByte: Byte; addr: Integer);
```
Funkce pro posílaní stavu finkcí F13-F28 a nastavení momentary/permanent.

petrilakm commented on src/tUltimateLI.pas:
```diff
@@ -604,6 +607,25 @@ procedure TuLI.ParseDeviceMsg(deviceAddr: Byte; var msg: TBuffer);
               Self.SendLocoData(msg.data[0], Self.LokAddrDecode(msg.data[3],
                 msg.data[4]));
             end;
+          07:
+            begin
+              Self.WriteLog(tllCommands, 'GET: locomotive Funktionsstatus F0-F12 request (>=3.6)');
+              Self.SendLocoFuncMomentary(msg.data[0], Self.LokAddrDecode(msg.data[3],
+                msg.data[4]));
+            end;
+          08:
+            begin
+              Self.WriteLog(tllCommands, 'GET: locomotive Funktionsstatus F13-F28 request (>=3.6)');
+              Self.SendLocoFunc13Momentary(msg.data[0], Self.LokAddrDecode(msg.data[3],
+                msg.data[4]));
+            end;
+          09:
+            begin
+              Self.WriteLog(tllCommands, 'GET: locomotive Funktionszustand F13-F28 request (>=3.6)');
+              Self.SendLocoFunc13(msg.data[0], Self.LokAddrDecode(msg.data[3],
+                msg.data[4]));
+            end;
+
```
Přidány zprávy dle specifikace 3.6 - vyčteno z dokumentu "XpressNET version 4.0"
V logu použíty originální nazvy zpráv, nepodařilo se mi to přeložit správně.
Funktionsstatus - zda je funkce momentary/permanent
Funktionszustand - zda je funkce zapnutá/vypnutá

petrilakm commented on src/tUltimateLI.pas:
```diff
@@ -113,6 +113,9 @@   TuLI = class
     procedure SendKeepAlive();
     procedure SendStatusRequest();
     procedure SendLocoData(callByte: Byte; addr: Integer);
+    procedure SendLocoFunc13(callByte: Byte; addr: Integer);
+    procedure SendLocoFuncMomentary(callByte: Byte; addr: Integer);
+    procedure SendLocoFunc13Momentary(callByte: Byte; addr: Integer);
```
Funkce pro posílaní stavu finkcí F13-F28 a nastavení momentary/permanent.

petrilakm commented on src/tUltimateLI.pas:
```diff
@@ -780,6 +802,52 @@ procedure TuLI.ParseDeviceMsg(deviceAddr: Byte; var msg: TBuffer);
+                (not Self.sloty[addr].isLoko)) then
+              begin
+                // lokomotiva neni rizena ovladacem
+                // -> odeslat "locomotive is being operated by another device"
+                Self.SendLokoStolen(Byte(msg.data[0]), Byte(msg.data[3]),
+                  Byte(msg.data[4]));
+              end
+              else
+              begin
+                // lokomotiva je rizena ovladacem -> nastavit funkce\
+                var funkce: TFunkce;
+                for var i := 0 to 7 do
+                  funkce[i + 13] := boolean((msg.data[5] shr i) and $1);
+                Self.sloty[addr].SetFunctions(13, 20, funkce);
+              end;
+            end;
+
+          $28:
+            begin
+              Self.WriteLog(tllCommands, 'GET: set F21-F28 (>=3.6)');
+
+              var addr := Self.LokAddrDecode(msg.data[3], msg.data[4]);
+              if ((addr = 0) or (addr > _SLOTS_CNT) or
+                (not Self.sloty[addr].isLoko)) then
+              begin
+                // lokomotiva neni rizena ovladacem
+                // -> odeslat "locomotive is being operated by another device"
+                Self.SendLokoStolen(Byte(msg.data[0]), Byte(msg.data[3]),
+                  Byte(msg.data[4]));
+              end
+              else
+              begin
+                // lokomotiva je rizena ovladacem -> nastavit funkce\
+                var funkce: TFunkce;
+                for var i := 0 to 7 do
+                  funkce[i + 21] := boolean((msg.data[5] shr i) and $1);
+                Self.sloty[addr].SetFunctions(21, 28, funkce);
+              end;
+            end;
+
```
Přidány zprávy dle specifikace 3.6 - vyčteno z dokumentu "XpressNET version 4.0"
Avšak kódy $23 a $28 neodpovídají pdf dokumentu, jsou zjíštěny experimentálně s ovladačem LH-100 v3.6

petrilakm commented on src/tUltimateLI.pas:
```diff
@@ -1198,6 +1266,216 @@ procedure TuLI.SendLocoData(callByte: Byte; addr: Integer);
+  end;
+
+  if ((addr = 0) or (addr > _SLOTS_CNT) or (not Self.sloty[addr].isLoko) or
+    (Self.sloty[addr].ukradeno)) then
+  begin
+    // lokomotiva neni rizena ovladacem
+    toSend := toSend + #$A + #$80 + #0 + #0;
+    if ((addr > 0) and (addr < _SLOTS_CNT)) then
+      for var i := 0 to _MAX_FUNC do
+        Self.sloty[addr].mausFunkce[i] := false;
+  end
+  else
+  begin
+    // Kennung (static)
+    toSend := toSend + #$51;
+
+    // F13 - F20
+    var tmp : Integer := 0;
+    for var i := 13 to 20 do
+      tmp := tmp + (Byte(Self.sloty[addr].funkceType[i]) shl (i - 13));
+    toSend := toSend + AnsiChar(tmp);
+
+    // F21 - F28
+    tmp := 0;
+    for var i := 21 to 28 do
+      tmp := tmp + (Byte(Self.sloty[addr].funkceType[i]) shl (i - 21));
+    toSend := toSend + AnsiChar(tmp);
+
+    // R (emulated)
+    toSend := toSend + #$03; // 28 speed steps
+
+    Self.sloty[addr].mausFunkce := Self.sloty[addr].funkce;
+  end;
+
+  Self.WriteLog(tllCommands, 'PUT: function F13-F28 momentary information');
+  Self.Send(CreateBuf(toSend));
+end;
+
+/// /////////////////////////////////////////////////////////////////////////////
+
```
Nové odpovědi dle XpressNET specifikace 3.6.
Naprosto hloupě okopírovaný kód. Funguje, ale chtšlo by to to nějak výrazně zkrátit. 95% kódu se opakuje.
Kontrola, zda není jiná adresa loko a spol. je pořád stejné. Liší se jen 4 poslední byte, co se odesílají.

malinajan commented:
Rád podporu dalších příkazů mergnu, jen by bylo dobré nějak zmírnit tu duplikaci kódu...

petrilakm commented:
DOplikaci se můžu pokusit odstranit. Zatím mám problém, že ten kód skoro vůbec nechápu. Ono se tma nějak řeší, když má stejný slot najednou jiný ovladač a když daný slot nemá nikdo, asi. Ale nerozumím ani způsobu detekce ani, co se s tím pak dělá.
Budu si to muset přečíst asi několikrát. Pak to ideálně hodit do nějaké fukce, co se jen zavolá.

petrilakm committed:
version 3.6 messages code deduplication

Deduplicated code for function F12-F28 (v3.6) handling.
Better behaviour for stop button on LH-100 hand controller.

petrilakm commented:
Tak jsem to trochu upravil, aby se kód tolik neopakoval.
MultiMaus by se měla chovat stejně jak předtím, tam jsem doufám nic neupravil.
Ovladač LH-100, pokud se mu nastaví zastavení jako STOP (DCC on), tak klávesou "St" lze uvolnit loko z olvadače.
Nouzové nastavení jedné loko se prování dle manualu k LH-100 - změna směru když jízdní stupně nejsou 0. Směr se nezmění, ale pošle se nouzové zastavení pro danou loko.

petrilakm committed:
new version 1.4.0.0

updated version number

malinajan commented on src/tUltimateLI.pas:
```diff
@@ -595,6 +601,46 @@ procedure TuLI.ParseDeviceMsg(deviceAddr: Byte; var msg: TBuffer);
+        Self.Send(CreateBuf(#$60 + _CMD_DCC_ON));
+        Sleep(50);
+        Self.Send(CreateBuf(#$60 + _CMD_DCC_ON));
+        Sleep(50);
```
Není s tímhle nějaký problém? Zaseknout program na 300 ms mi připadá jako docela dlouhá doba. Nebylo by lepší implementovat nějaký timer nebo tak něco?

petrilakm replied:
Jaj, tohle bylo řešení jen na rychlo, zda to pomůže. A pomohlo to. Ano správně se to má řešič přes timer. Teď to dej klidně pryč. Já to pak můžu vyřešit lépe nějakým dalším commitem.

malinajan reviewed (changes requested):
Pročetl jsem si tvůj PR na uli-daemon a vyrobil jsem nad tvými commity další víceméně kosmetické commity. Pak mám jednu větší připomínku (viz konkrétní komentář v kódu). Myslím, že stojí za to, aby sis moje commity alespoň prolítl a aby byly v historii PR na GitHubu. A hlavně, pokud budeš řešit onu připomínku, aby bylo její řešení už i nad mými commity. Přemýšlím, jak to udělat.

Asi nejlepší mi přijde, abys mi dal přístup do tvého forku a já tam prostě svoje kosmetické commity přidám. Nebo můžu commity pushnout do nějaké náhodné větve v `kmzbrnoI/uli-daemon` a ty si je do svého forku pullneš sám. Sám nevím, co je lepší. Máš nějakou preferenci?

malinajan commented:
Moje drobné úpravy nad tvými jsou ve větvi `mp-master`: <https://github.com/kmzbrnoI/uLI-daemon/tree/mp-master>. Pullni si je k sobě, jestli s nimi nemáš problém. Mám s mergnutím tohoto PR počkat, než vyřešíš to časování?
